### PR TITLE
Deprecate dms_to_degrees

### LINKED ISSUE #12239: Angle bug for (d, m, s) tuple input (deprecate dms_to_degrees)
`Angle` does not handle the sign correctly for a `(d, m, s)` tuple input if `d=0`:

```python
>>> from astropy.coordinates import Angle
>>> ang = Angle((-0, -42, -17), unit='deg')
>>> print(ang)
0d42m17s
>>> print(ang.dms)
dms_tuple(d=0.0, m=42.0, s=16.999999999999886)
>>> print(ang.signed_dms)
signed_dms_tuple(sign=1.0, d=0.0, m=42.0, s=16.999999999999886)
```

### Expected behavior

```python
>>> ang = Angle((-0, -42, -17), unit='deg')
>>> print(ang)
-0d42m17s
>>> print(ang.dms)
dms_tuple(d=-0.0, m=-42.0, s=-16.999999999999886)
>>> print(ang.signed_dms)
signed_dms_tuple(sign=-1.0, d=0.0, m=42.0, s=16.999999999999886)
```

### PR DESCRIPTION
**Pull request by eteq**
### Description

This PR is essentially a replacement for #12443 and fixes #12239 - in its current state it simply adds a deprecation warning for ``dms_to_degrees`` a la https://github.com/astropy/astropy/issues/12239#issuecomment-964345806, - I agree with @mhvk's point that this is more trouble than it's worth to really try to work out the user's intent. The deprecation warning now specifically says essentially "convert to degrees yourself, or use the string input where there's not sign ambiguity".

I did explore a few other options, though, and so there's a few extraneous commits.  That is *intentional*, so I do not want to squash it out, as it shows how a follow-on PR might address this use case if we wanted to.  Right now I just want to make sure we get the deprecation warning in for 5.1 since deprecating sooner is always better.

(Note this also contains the commits #12443 - that is also intentional because I don't think we should merge #12443 but @webbdays deserves some credit for sleuthing out some of these corner cases - if someone wants I am fine squashing all that down to one commit but :shrug: to me it didn't seem worth the effort).

EDIT: Close #12443

### Checklist for package maintainer(s)

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

### CODE REVIEW

**Comment by pllim** [comment]
Maybe squash into one commit and give @webbdays co-author credit?

https://docs.github.com/en/enterprise-cloud@latest/pull-requests/committing-changes-to-your-project/creating-and-editing-commits/creating-a-commit-with-multiple-authors

**Comment by pllim** [comment]
12 commits for 8 lines seems a little excessive.

**Comment by pllim** [comment]
Also

```
fix end of files.........................................................Failed
- hook id: end-of-file-fixer
- duration: 0.07s
- exit code: 1
- files were modified by this hook
```

File: astropy/coordinates/angle_formats.py
Comment (pllim) [inline astropy/coordinates/angle_formats.py:420]: Why not use the `@deprecated` decorator?

File: astropy/coordinates/angle_formats.py
Comment (pllim) [inline astropy/coordinates/angle_formats.py:416]: I don't think this should be indented.
```suggestion
    """
```

**Comment by eteq** [comment]
> Maybe squash into one commit and give @webbdays co-author credit?
> 12 commits for 8 lines seems a little excessive.

I disagree!  I don't think that reflects what actually happened here (i.e. @webbdays didn't co-author my commits), and in my view the purpose of the commit log is to record the path of development, which is exactly what this does.  Sometimes a single line of code has embedded in it hours, days, or years of discussion (especially in `coordinates`...), and this is a nod to that concept.  Of course 1 commit also doesn't equal 1 unit of work, but it seems better to record something rather than nothing.

(As I said above, though, I am perfectly willing to squash @webbdays's commits into just 1 given that "a commit" maps onto "unspecified number of thoughts".)

(Will fix the typos momentarily!)

File: docs/changes/coordinates/13162.api.rst
Comment (mhvk) [inline docs/changes/coordinates/13162.api.rst:1]: Let's add a sentence of *why* it is deprecated. Something like ", as it was difficult to be sure about the intent of the user for signed values of the degrees, minutes, and seconds."

**Review by mhvk — COMMENTED** [review]
Comment about the changelog entry beyond the ones by @pllim. I agree with her that we should just squash everything together (ie., joint commit); I doubt we really want to get back to this, and if we do, finding the github issue is likely easier than trawling through the history of the file.

File: astropy/coordinates/angle_formats.py
Comment (eteq) [reply astropy/coordinates/angle_formats.py:420]: Hmm, at the time my motivation was because there's not a single clear "alternative" to use, but now that I look at how ``alternative`` is rendered I think I can make it work.  One sec.

File: docs/changes/coordinates/13162.api.rst
Comment (eteq) [reply docs/changes/coordinates/13162.api.rst]: Done!

**Comment by mhvk** [comment]
@eteq - not really worth too much trouble, but the practice in the linux kernel seems to be to use one commit per functional step.  But then, Linus also insists on commit entries that are actually meaningful, and contain thoughts about implementation, etc., which github is extremely poorly set up to deal with. Instead, it is all in the conversation, which is not part of the repository's history (so that one is locked in, at least to some extend).

Anyway, in this particular case I still see very little of value in the various commits, but, heck, not going to stop progress for it!

**Comment by pllim** [comment]
Huh, no test needs to be updated to catch this new deprecation warning? Is this not tested?

I still prefer squashing... 👻

**Comment by eteq** [comment]
@mhvk - Yeah, I know that philosophy (although did not know that was how things go in the linux kernel).  I just think it's wrong: To me part of the brilliance of `git` is that commits are really cheap, and so it's *good* and *proper* to use them to store information about the development history, and it's also right to say that this is different for different people, since its purpose is to encode thinking rather than the code per se. (Caveat: if it leads to performance issues or makes bisect a lot more annoying, it's not worth it, but I think our practice of no complicated merges without rebase addresses the latter better anyway, and the former is not an issue here because the commits are very small).

That said, I am clearly outvoted in this thread, if neither of you are convinced by the above.  Just know that when the code archeologists from 1000 years in the future come to this thread, they'll say "boy I wish they'd listened to @eteq" 😉. I'll squash it down to just one commit from me and one from @webbdays.

**Comment by pllim** [comment]
FWIW the original commits are still in #12443 for archeologists as long as Microsoft do another Arctic Vault backup.

**Comment by eteq** [comment]
(lets see what the CI does now @pllim - I'm also a bit weirded out by the fact that none of the tests exercised this.  But if so it just supports @mhvk's point it is a little-used feature!)

**Review by pllim — APPROVED** [review]
LGTM. Thanks!

Let's wait for @mhvk to have one more look as well, if time allows.

**Comment by eteq** [comment]
Also, while leaving notes for posterity, if in the future someone really wants to keep this function, https://github.com/eteq/astropy/commit/82fa4dc9755d7f568f854bdc75902d797d2d8dcf has the start I squashed out of this which can be a starting point for a cleaner implementation

**Review by mhvk — APPROVED** [review]
Happy with this (assuming CI passes -- which would suggest the function was indeed not tested...)

**Comment by pllim** [comment]
4 failures... so it is tested after all.

**Comment by eteq** [comment]
Aha, it turns out this function is used very indirectly internally when parsing strings like "1d2m3.4s" which makes sense in retrospect.  Last commit here now supresses the deprecation in those cases, because they are *already* validated/tested for the case "-1d2m3.4s", and an error is raised for the string "1d-2m3.4s" or similar.  

We should probably add in a new issue once this is merged that makes sure we don't lose the thread around potentially cleaning up or totally eliminating degrees_to_dms, but that should be after this is merged.

File: astropy/coordinates/angles.py
Comment (pllim) [inline astropy/coordinates/angles.py:155]: I don't know how I feel about this. If we are serious about deprecating this, we should not even use it internally.

If we have no choice but to use it internally, then instead of deprecating, we should fix it properly.

File: astropy/coordinates/angles.py
Comment (mhvk) [inline astropy/coordinates/angles.py:128]: This suppression of the deprecation seems weird. Wouldn't it be better to just have a direct implementation that does the right thing? That way, once the deprecation period is over, we can just delete `dms_to_degrees` and `_tuple_to_float`. 

Here, because the parser has already ensured the various elements are of the right type, I think it is as simple as 
```
b, m, s = angle
if angle_unit == u.hr:
    form._check_hour_range(b)
form._check_minute_range(m)
form._check_second_range(s)
angle = np.copysign(np.abs(b]) + m / 60. + s / 3600., b) << angle_unit
```

**Review by mhvk — COMMENTED** [review]
The new commit is a bit weird (see in-line comment). It also makes me realize that tuple input for hourangle is still allowed. That seems a bit strange!

File: astropy/coordinates/angles.py
Comment (eteq) [reply astropy/coordinates/angles.py:126]: Fair point - I was trying to do the minimal possible, but this is a better idea!

Comment (eteq) [reply astropy/coordinates/angles.py:126]: Hmm.  On second thought, when I went to implement this I got a big piles of errors.  It's partly because the tuples are not constrained to be length-3, but even when I account for that something else is broken.

So I think we should make it a project for 5.2 to do this correctly, but not have it hold up this PR which is already going to need special permission to make it into 5.1!

**Comment by eteq** [comment]
@mhvk - I agree with you the suppressing of the deprecation is rather hacky, but the trivial re-implementation is not working for unclear reasons (see above).  So I think if we want this deprecation in for 5.1 we should leave it as is and sort it out for 5.2 (but if you're not comfortable with that, we can also just delay it to 5.2...)

**Comment by pllim** [comment]
My vote is to just delay the deprecation itself to v5.2. It does not feel right to deprecate something that we are actively using without a clear replacement.

**Comment by mhvk** [comment]
Hmm, I think we indeed might as well delay to 5.2. The path forward at least seems fairly clear (even if it is trickier than thought to implement...)

**Comment by pllim** [comment]
I am converting this to draft until the implementation issues are resolved. Thanks!

**Comment by mhvk** [comment]
@eteq - I thought we should try to finish this so adjusted my solution suggested above to make it work. But then I also realized we logically should deprecate `hms_to_hours` as well (same problem, just less often obvious)...  So, I did that too and force-pushed to your branch. I *think* it is now all OK.

**Comment by mhvk** [comment]
p.s. I put back the backport label, but obviously we don't strictly have to do that.

**Comment by pllim** [comment]
@meeseeksdev backport to v5.1.x

**Comment by pllim** [comment]
Thanks! I opened #13179 as a reminder to remove them.

**Pull request merged**
